Introduce UnordMap, UnordSet, and UnordBag (MCP 533)

michaelwoerister opened the pull request:
This is the start of implementing [MCP 533](https://github.com/rust-lang/compiler-team/issues/533).

I followed @eddyb's suggestion of naming the collection types `Unord(Map/Set/Bag)` which is a bit easier to type than `Unordered(Map/Set/Bag)`

r? @eddyb

bors commented:
:umbrella: The latest upstream changes (presumably #103228) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3330154739/jobs/5508271130) [(plain)](https://github.com/rust-lang/rust/commit/8d7f17fd1244eec9ecff46740cae0895a9369849/checks/9120231230/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
running 24 tests
.......i.......F........
failures:

---- src/unord.rs - unord::UnordItems (line 25) stdout ----
error[E0425]: cannot find value `unordered_items` in this scope
 --> src/unord.rs:27:1
  |
4 | unordered_items.all(|x| ordered.push(x));

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
For more information about this error, try `rustc --explain E0425`.
Couldn't compile the test.

failures:
    src/unord.rs - unord::UnordItems (line 25)
test result: FAILED. 22 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out; finished in 0.57s


error: doctest failed, to rerun pass `-p rustc_data_structures --doc`
```

</details>

michaelwoerister committed:
Introduce UnordMap, UnordSet, and UnordBag (see MCP 533)

MCP 533: https://github.com/rust-lang/compiler-team/issues/533

Also, as an example, substitute UnordMap for FxHashMap in
used_trait_imports query result.

michaelwoerister commented:
> r? compiler

Hm, that didn't seem to work...

michaelwoerister commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 9117ea975834a86dadcb9ebbc40dd9a9fb0f78ae with merge 8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9117ea975834a86dadcb9ebbc40dd9a9fb0f78ae","merge_sha":"8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c"} -->

lcnr commented:
r? @lcnr

lcnr commented:
r=me after perf. Is the end goal here to completely remove `FxHashSet/Map` from the compiler?

michaelwoerister commented:
The MCP only states the goal of removing the `HashStable` implementations, which means that they can't be used in query keys or results anymore. But I'd be fine with removing them altogether.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3343958233/jobs/5539593977)
Build commit: 8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c (`8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3343958233/jobs/5539593977"},"merge_sha":"8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3343958233/jobs/5539593977)
Build commit: 8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c (`8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3343958233/jobs/5539593977"},"merge_sha":"8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c"} -->

rust-timer commented:
Queued 8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c with parent cdd7afeaadf1c48eafb4dff4452439fa5d13a775, future [comparison URL](https://perf.rust-lang.org/compare.html?start=cdd7afeaadf1c48eafb4dff4452439fa5d13a775&end=8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c).

rust-timer commented:
Finished benchmarking commit (8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c): [comparison URL](https://perf.rust-lang.org/compare.html?start=cdd7afeaadf1c48eafb4dff4452439fa5d13a775&end=8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=cdd7afeaadf1c48eafb4dff4452439fa5d13a775&end=8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 0.4%     | [0.4%, 0.4%] | 2         |
| Improvements ✅ <br /> (primary) | -0.7%    | [-0.7%, -0.7%] | 1         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -0.7%    | [-0.7%, -0.7%] | 1         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=cdd7afeaadf1c48eafb4dff4452439fa5d13a775&end=8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=cdd7afeaadf1c48eafb4dff4452439fa5d13a775&end=8f2030cfc7fe4c36f7e41a7593c5d23356c06f5c&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

michaelwoerister commented:
Perf looks unchanged -- I don't think the two minor changes are real.

@bors r=lncr

bors commented:
:pushpin: Commit 9117ea975834a86dadcb9ebbc40dd9a9fb0f78ae has been approved by `lncr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lncr 9117ea975834a86dadcb9ebbc40dd9a9fb0f78ae -->
<!-- homu: {"type":"Approved","sha":"9117ea975834a86dadcb9ebbc40dd9a9fb0f78ae","approver":"lncr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

michaelwoerister commented:
Thanks for the review, @lncr!

bors commented:
:hourglass: Testing commit 9117ea975834a86dadcb9ebbc40dd9a9fb0f78ae with merge 607878d069267e1402ad792c9331b426e4c6d0f9...
<!-- homu: {"type":"BuildStarted","head_sha":"9117ea975834a86dadcb9ebbc40dd9a9fb0f78ae","merge_sha":"607878d069267e1402ad792c9331b426e4c6d0f9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3350367972/jobs/5551985884)
Approved by: lncr
Pushing 607878d069267e1402ad792c9331b426e4c6d0f9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lncr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3350367972/jobs/5551985884"},"merge_sha":"607878d069267e1402ad792c9331b426e4c6d0f9"} -->

rust-timer commented:
Finished benchmarking commit (607878d069267e1402ad792c9331b426e4c6d0f9): [comparison URL](https://perf.rust-lang.org/compare.html?start=33b55ac39fa633d0983fad014469e1036669bf28&end=607878d069267e1402ad792c9331b426e4c6d0f9&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=33b55ac39fa633d0983fad014469e1036669bf28&end=607878d069267e1402ad792c9331b426e4c6d0f9&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3% | [1.2%, 1.4%] | 2     |
| Regressions ❌ <br /> (secondary)  | 1.8% | [0.4%, 4.1%] | 14    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.3% | [1.2%, 1.4%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=33b55ac39fa633d0983fad014469e1036669bf28&end=607878d069267e1402ad792c9331b426e4c6d0f9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.8% | [3.5%, 4.2%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=33b55ac39fa633d0983fad014469e1036669bf28&end=607878d069267e1402ad792c9331b426e4c6d0f9&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.0%, 2.3%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

lqd commented:
These benchmarks are currently noisy so I think we're good here, @rustbot label: +perf-regression-triaged

![image](https://user-images.githubusercontent.com/247183/198827464-97a706e6-842a-4320-ad68-152c4348579a.png)
